### PR TITLE
Performance optimization across the full engine

### PR DESCRIPTION
**Pull request by DrRataplan**
In a nutshell:

- Statically evaluate some expressions: note that function calls are never evaluated statically, this will be done in the future
- Cache casting transforms, and optimize them
- Optimize value compares by skipping the whole type system: Javascript knows no difference between float, double, decimal, in, etc. Plainly comparing them makes stuff faster.
- Re-use boolean instances, and make a boolean-singleton-sequence, to improve memory usage.

### CODE REVIEW

File: test/unrunnableTestCases.csv [hunk 1]
```diff
@@ -2103,7 +2103,7 @@ CastAs673b,Error: XPTY0004: Casting not supported from xs:string to xs:QName.
 CastAs677,Error: XPST0017: Function string-to-codepoints with arity of 1 not registered. No similar functions found.
 CastAs678,Error: XPST0017: Function string-to-codepoints with arity of 1 not registered. No similar functions found.
 CastFOCA0003-1,AssertionError: Expected executing the XPath "xs:integer(99e100)" to resolve to one of the expected results, but got Error: FOAR0002: can not cast 9.9e+101 to xs:integer, it is out of bounds for JavaScript numbers., AssertionError: expected [Function] to throw error including 'FOCA0003' but got 'FOAR0002: can not cast 9.9e+101 to xs:integer, it is out of bounds for JavaScript numbers.'.
-K-SeqExprCast-8,AssertionError: expected [Function] to throw error including 'XPST0081' but got 'Cannot read property \'na\' of undefined'
+K-SeqExprCast-8,AssertionError: expected [Function] to throw error including 'XPST0081' but got 'Cannot read property \'pa\' of undefined'
```
Comment (bwrrp) [inline test/unrunnableTestCases.csv]: make this throw a nice error?

File: src/evaluateXPath.js [hunk 2]
```diff
@@ -27,9 +28,11 @@ import isInstanceOfType from './selectors/dataTypes/isInstanceOfType';
...
  */
 function evaluateXPath (xpathSelector, contextItem, domFacade, variables = {}, returnType = evaluateXPath.ANY_TYPE, options = {}) {
-	if (!xpathSelector || typeof xpathSelector !== 'string' ) {
-		throw new TypeError('Failed to execute \'evaluateXPath\': xpathSelector must be a string.');
-	}
+	(function checkParameters () {
+		if (!xpathSelector || typeof xpathSelector !== 'string' ) {
+			throw new TypeError('Failed to execute \'evaluateXPath\': xpathSelector must be a string.');
+		}
+	})();
```
Comment (bwrrp) [inline src/evaluateXPath.js]: is this IIFE necessary?

Comment (bwrrp) [inline src/evaluateXPath.js]: ...same for the others

File: src/selectors/DynamicContext.js [hunk 3]
```diff
@@ -52,10 +52,42 @@ class DynamicContext {
 	createScopedContext (overlayContext) {
```
Comment (bwrrp) [inline src/selectors/DynamicContext.js]: Remove entirely / replace with scopeWithEverything (without the !== undefined ternaries)

File: src/selectors/axes/DescendantAxis.js [hunk 4]
```diff
@@ -1,6 +1,28 @@
...
+
+LinkedListIterator.prototype.next = function () {
+    const head = this.head;
+    if (!head) {
+        return { done: true, value: undefined };
+	}
+
+    const children = this._domFacade.getChildNodes(head.node);
+    for (let i = 0, len = children.length, r = head; i < len; ++i) {
+        r = r.rest = { node: children[i], rest: r.rest };
```
Comment (bwrrp) [inline src/selectors/axes/DescendantAxis.js]: o_O

File: src/selectors/dataTypes/casting/castToDuration.js [hunk 5]
```diff
@@ -4,33 +4,34 @@ import Duration from '../valueTypes/Duration';
...
+		return (value) => ({
 			successful: true,
 			value: createDurationValue(value)
-		};
+		});
 	}
 	if (instanceOf('xs:untypedAtomic') || instanceOf('xs:string')) {
-		const parsedDuration = Duration.fromString(value, 'xs:duration');
-		if (parsedDuration) {
+		return (value) => {
```
Comment (bwrrp) [inline src/selectors/dataTypes/casting/castToDuration.js]: formatting is not consistent: (value) => { ... } or value => { ... }

File: src/selectors/dataTypes/casting/castToFloat.js [hunk 6]
```diff
@@ -2,17 +2,22 @@ import castToFloatLikeType from './castToFloatLikeType';
...
-	return {
-		successful: true,
-		value: createAtomicValue(castResult.value, 'xs:float')
+export default function castToFloat (instanceOf) {
+	/**
+	 * @type {function (./AtomicValueDataType) : ({successful: boolean, value: ../AtomicValue}|{successful: boolean, error: !Error})}
+	 */
+	const caster = castToFloatLikeType(instanceOf, 'xs:float');
+	return value => {
+		const castResult = caster.call(undefined, value);
```
Comment (bwrrp) [inline src/selectors/dataTypes/casting/castToFloat.js]: caster.call(undefined, value) is the same as caster(value)
Only calling from an expression with a . changes the this value.

File: test/unrunnableTestCases.csv (hunk 1 above)
Comment (DrRataplan) [inline test/unrunnableTestCases.csv]: The test is 
`   <test-case name="K-SeqExprCast-8">
      <description> A prefix is used which isn't declared.</description>
      <created by="Frans Englich" on="2007-11-26"/>
      <test>3 cast as prefixDoesNotExist:integer</test>
      <result>
         <error code="XPST0081"/>
      </result>
   </test-case>
`

I'll make an issue for it

**Pull request merged**